### ISSUE TITLE
Foreground service doesn't work in Android 10

### ISSUE THREAD

**Issue by JonasWestAlro** [mauron85/cordova-plugin-background-geolocation#691]
Hi, 

In Android 10 the service definition in the manifest file must have: `android:foregroundServiceType="location"` in order for the foreground service to work properly (ie. when user select "only while using app" as location permission. 

` <service android:enabled="true" android:exported="false" android:foregroundServiceType="location" android:name="com.marianhello.bgloc.service.LocationServiceImpl" />
`

I hope it is possible to add this in the next version. 

Does anyone have a quick fix for this? My current solution is to add the line manually to the manifest file after each "cordova prepare". It is just a matter of time before I forget it before releasing.

**Comment by hatfield115** [comment]
I had this same issue and just added this to the config.xml file (within the platform name=android section)

        <edit-config file="AndroidManifest.xml" mode="merge" target="/manifest/uses-permission" xmlns:android="http://schemas.android.com/apk/res/android">
            <uses-permission android:name="android.permission.FOREGROUND_SERVICE" />
        </edit-config>

**Comment by betagole** [comment]
I added this to my config.xml and 'android:foregroundServiceType="location"' still doesn't appear on my  <service android:enabled="true" android:exported="false" android:name="com.marianhello.bgloc.service.LocationServiceImpl" /> tag.

**Comment by hatfield115** [comment]
You may want  to run prepare and make sure it regenerates the file, if that still doesnt work, send me the full config file and I will see if anything jumps out that could be causing it to not work.

**Comment by betagole** [comment]
[config.txt](https://github.com/mauron85/cordova-plugin-background-geolocation/files/4546538/config.txt)

**Comment by hatfield115** [comment]
Not sure why its not working - Im assuming you removed the current file just to make sure it recreated a new one?  Other than that the file path is the only difference I see - but if your others are working then I would presume this one should to.  Just to rule it out, you may want to use what I have exactly - filename and the XMLNS:android (I know you have it in the header - but again - just to rule it out.

Good luck.

**Comment by betagole** [comment]
When you say remove the current file, do you mean '~platforms\android\app\src\main\AndroidManifest.xml'? I removed that and ran prepare and it did not recreate it, it errored out trying to find it.

**Comment by spinninghamster** [comment]
Anyone got this working?

**Comment by hatfield115** [comment]
I did... Had no issues.  What problem are you seeing?

**Comment by mobilemarines** [comment]
Hi guys,
as Android SDK 29 is mandatory till November 2, 2020 this is a quite urgent issue to keep this plugin alive.

I made the geoposition updates work in background (without ALWAYS permission) with these adjustments:
Angular:
```
this.backgroundGeolocation.configure({
	startForeground: true,
	activityType: "AutomotiveNavigation",
	saveBatteryOnBackground: false
});
```

AndroidManifest.xml:
```
<service android:enabled="true" android:exported="false" android:foregroundServiceType="location" android:name="com.marianhello.bgloc.service.LocationServiceImpl" />
<uses-permission android:name="android.permission.REQUEST_IGNORE_BATTERY_OPTIMIZATIONS" />
<uses-permission android:name="android.permission.ACCESS_BACKGROUND_LOCATION" />
```

If you do that manually android.json must be adjusted the same way.
It would be best to adjust the plugin accordingly so the correct permissions are automatically rendered in all corresponding config files.

The remaining problem I have is that the position updates are not sent in background (after 5 minutes).
I guess there is missing android:foregroundServiceType="dataSync" at the corresponding service (com.marianhello.bgloc.sync.SyncService?).
I also tried to add 'location|dataSync' to both but still http(s) posts in background do not work, but they are postoned until the app is back in foreground.
This was working even separately (after receiving a GPS position) with the AdvancedHttp plugin in background with target SDK 28.

Does anyone know how to make that work?
I use the npm plugin version 3.1.0, Cordova 9, Ionic 4, Angular 7.

**Comment by isinocasio** [comment]
> Hi guys,
> as Android SDK 29 is mandatory till November 2, 2020 this is a quite urgent issue to keep this plugin alive.
> 
> I made the geoposition updates work in background (without ALWAYS permission) with these adjustments:
> Angular:
> 
> ```
> this.backgroundGeolocation.configure({
> 	startForeground: true,
> 	activityType: "AutomotiveNavigation",
> 	saveBatteryOnBackground: false
> });
> ```
> 
> AndroidManifest.xml:
> 
> ```
> <service android:enabled="true" android:exported="false" android:foregroundServiceType="location|dataSync" android:name="com.marianhello.bgloc.service.LocationServiceImpl" />
> <service android:exported="true" android:foregroundServiceType="location|dataSync" android:name="com.marianhello.bgloc.sync.SyncService" android:process=":sync">
> 	<intent-filter>
> 		<action android:name="android.content.SyncAdapter" />
> 	</intent-filter>
> 	<meta-data android:name="android.content.SyncAdapter" android:resource="@xml/syncadapter" />
> </service>
> <uses-permission android:name="android.permission.REQUEST_IGNORE_BATTERY_OPTIMIZATIONS" />
> ```
> 
> If you do that manually android.json must be adjusted the same way.
> It would be best to adjust the plugin accordingly so the correct permissions are automatically rendered in all corresponding config files.

Idid just as you told with plugin.xml and android.json in the plugin folder but project mainActivity seems to desapear (dont know why), and throws "No Java files found that extend CordovaActivity."

**Comment by chacabuk** [comment]
Can somebody make it work on Android 10 without  ALWAYS permission?

**Comment by isinocasio** [comment]
> Can somebody make it work on Android 10 without ALWAYS permission?

I used this fork of this same plugin:
cordova plugin add https://github.com/Astroxslurg/cordova-plugin-background-geolocation.git

That has all the necesary changes on manifest that makes this plugin work on foreground.
Works like a charm, try it out.
Dont know why the oficial repository does not have this necessary changes though.

**Comment by mobilemarines** [comment]
To be honest, this (discontinued) plugin and all (I guess - don't know all) forks are dying due to the lack of adjustments to the latest SDKs and other evolution.
There have been a lot of changes in Android 10 + 11 regarding the background (always) permission and capability of sending position updates in the background. Also iOS 14 introduced the approximate location.
Additionally many major vendors kill apps in the background, see https://dontkillmyapp.com/
So this plugin should actually be renamed to cordova-plugin-foreground-geolocation. ;-)
Only the original (paid) plugin seems to keep track of the plugin to work in the background/minimized: https://github.com/transistorsoft/cordova-background-geolocation-lt

**Comment by isinocasio** [comment]
> To be honest, this (discontinued) plugin and all (I guess - don't know all) forks are dying due to the lack of adjustments to the latest SDKs and other evolution.
> There have been a lot of changes in Android 10 + 11 regarding the background (always) permission and capability of sending position updates in the background. Also iOS 14 introduced the approximate location.
> Additionally many major vendors kill apps in the background, see https://dontkillmyapp.com/
> So this plugin should actually be renamed to cordova-plugin-foreground-geolocation. ;-)
> Only the original (paid) plugin seems to keep track of the plugin to work in the background/minimized: https://github.com/transistorsoft/cordova-background-geolocation-lt

That is true, actually Im starting to train myself in Flutter because all of this issues with cordova.
